manual/grammar/grammar.ebnf

```diff
@@ -0,0 +1,188 @@
+  delete_state  ::=  >('delete' _term (';' _process)?)
+  read_state  ::=  >LOOKUP('lookup' (_term: from) 'as' (_lvar: to) 'in' (_process: in) ('else' (_process: else))? (';' _process)?)
+  set_lock  ::=  >('lock' _term (';' _process)?)
+  remove_lock  ::=  >('unlock' _term (';' _process)?)
+  _condition  ::=  equality_check | lesser_check | predicate_ref
+  equality_check  ::=  (_term | _formula) @(1('=')) (_term | _formula)
+  lesser_check  ::=  _term ('(<)' | '<<') _term
+  let  ::=  'let' (_term: let_identifier) '=' _process
+  export  ::=  'export' (ident: export_identifier) ':' '"' export_query '"'
+  _rule  ::=  rule | diff_rule
+  rule  ::=  simple_rule variants?
+  diff_rule  ::=  simple_rule 'left' (rule: left) 'right' (rule: right)
+  simple_rule  ::=  'rule' modulo? (ident: rule_identifier) rule_attrs? ':' rule_let_block? premise ('-->' | action_fact) conclusion
+  premise  ::=  '[' _facts? ']'
+  action_fact  ::=  '--[' _facts_restrictions? ']->'
+  conclusion  ::=  '[' _facts? ']'
+  variants  ::=  'variants' simple_rule (',' simple_rule)*
+  modulo  ::=  '(' 'modulo' ('E' | 'AC') ')'
+  rule_attrs  ::=  '[' rule_attr (',' rule_attr)* ','? ']'
+  rule_attr  ::=  ('color=' | 'colour=') hexcolor
+  rule_let_block  ::=  'let' rule_let_term+ 'in'
+  rule_let_term  ::=  ((msg_var_or_nullary_fun | nat_var): left) '=' (_term: right)
+  macros  ::=  'macros' ':' macro (',' macro)*
+  macro  ::=  (ident: macro_identifier) '(' (_non_temporal_var (',' _non_temporal_var)*)? ')' '=' (_term: term)
+  embedded_restriction  ::=  '_restrict' '(' (_formula: formula) ')'
+  _facts_restrictions  ::=  <((_fact | embedded_restriction) (',' (_fact | embedded_restriction))*)
+  _facts  ::=  <(_fact (',' _fact)*)
+  _fact  ::=  (fact -> linear_fact) | ('!' (fact -> persistent_fact))
+  fact  ::=  <((ident: fact_identifier) '(' arguments? ')' fact_annotes?)
+  fact_annotes  ::=  '[' fact_annote (',' fact_annote)* ']'
+  fact_annote  ::=  '+' | '-' | 'no_precomp'
+  restriction  ::=  ('restriction' | 'axiom') (ident: restriction_identifier) restriction_attr? ':' '"' (_formula: formula) '"'
+  restriction_attr  ::=  '[' ('left' | 'right') ']'
+  case_test  ::=  'test' (ident: test_identifier) ':' '"' (_formula: formula) '"'
+  _lemma  ::=  lemma | diff_lemma | accountability_lemma | equiv_lemma | diff_equiv_lemma
+  lemma  ::=  'lemma' modulo? (ident: lemma_identifier) lemma_attrs? ':' trace_quantifier? '"' (_formula: formula) '"' (_proof_skeleton: proof_skeleton)?
+  lemma_attrs  ::=  '[' lemma_attr (',' lemma_attr)* ','? ']'
+  lemma_attr  ::=  'sources' | 'reuse' | 'use_induction' | ('output=' '[' language (',' language)* ']') | ('hide_lemma=' ident) | ('heuristic=' (_goal_ranking+: goal_ranking))
+  language  ::=  'spthy' | 'spthytyped' | 'msr' | 'proverif' | 'deepsec'
+  diff_lemma  ::=  ('diffLemma' | 'lemma') modulo? (ident: lemma_identifier) diff_lemma_attrs? ':' trace_quantifier? ('"' (_formula: formula) '"')? (_proof_skeleton: proof_skeleton)?
```

Comment:
Why does that case `diff_lemma` also include a normal lemma, although this is already covered above in the case `lemma` above? This leads to a situation where a normal lemma is sometimes parsed as a `diff_lemma`.

Reply:
In that case we can also remove the trace quantifier and the formula.

Reply:
Hi! I've tried that. In `examples/ccs18-5G/5G-AKA-untraceability/proofs/5G_AKA_passive_privacy_game__allProofsExcept-Equ-XOR-and-Equality.spthy`, we have

```
lemma dummy [right]:
  all-traces "¬(⊥)"
/*
```

Do you want this to be parsed as a lemma (then lemmas can have `right` and `left` as attributes, I suppose?) or as a `diff_lemma`?

Reply:
As a lemma. diff_lemma should be reserved for the observational equivalence lemma.